torch/_inductor/triton_bundler.py

```diff
@@ -0,0 +1,217 @@
+                                )
+                            counters["inductor"]["triton_bundler_save_kernel"] += 1
+                        except Exception:
+                            log.debug("failed to collect triton kernel", exc_info=True)
+                        if filename.endswith(".cubin"):
+                            # Each kernel has bunch of files like .cubin, .json, .ttir
+                            # Just append one of them without the extension
+                            kernel_names.append(Path(filename).stem)
+                    if artifacts:
+                        result.append(
+                            TritonKernelArtifacts(
+                                entry.kernel_hash,
+                                entry.device,
+                                artifacts,
+                            )
+                        )
+                cls._entries = None
+                return result, TritonBundlerMetadata(kernel_names)
+            return [], None
+
+    @staticmethod
+    def read_and_emit(
+        bundle: List[TritonKernelArtifacts],
+    ) -> Optional[TritonBundlerMetadata]:
+        """
+        This is the main function called when a cache read happens. This function
+        converts the bundled format back into individual files and writes them
+        to the filesystem.
+        """
+        if not TritonBundler.is_enabled():
+            return None
+
+        with dynamo_timed(key="TritonBundler.read_and_emit", fwd_only=False):
+            kernel_names: List[str] = []
+
+            for artifacts in bundle:
+                basedir = triton_cache_dir(artifacts.device)
+                directory = os.path.join(basedir, artifacts.kernel_hash)
+                Path(directory).mkdir(parents=True, exist_ok=True)
+
+                # Random ID to avoid any collisions
+                rnd_id = str(uuid.uuid4())
+                tmp_dir = os.path.join(basedir, f"tmp.{rnd_id}")
+                os.makedirs(tmp_dir)
+
+                for artifact in artifacts.artifacts:
+                    filepath = os.path.join(tmp_dir, artifact.filename)
+                    with open(filepath, "wb") as file:
+                        payload = artifact.payload
+                        if artifact.filename.endswith(".json"):
+                            payload = payload.replace(
+                                TritonBundler._REPLACE_BYTES, str.encode(directory)
+                            )
+                        file.write(payload)
+                    counters["inductor"]["triton_bundler_read_and_emit_kernel"] += 1
+                    if artifact.filename.endswith(".cubin"):
+                        # Each kernel has bunch of files like .cubin, .json, .ttir
+                        # Just append one of them without the extension
+                        kernel_names.append(Path(artifact.filename).stem)
+                # Atomic on POSIX systems
+                os.replace(tmp_dir, directory)
```

Comment:
A little confused here:

> os.replace(src, dst, *, src_dir_fd=None, dst_dir_fd=None)
> Rename the file or directory src to dst. If dst is a non-empty directory, OSError will be raised.

You seem to pass a directory on the RHS, this doesn't seem like it could possibly work?

Reply:
we chatted offline, this does work